Improve KMS documentation (#493)
digital-asset/docs.daml.com: docs/2.8.0/docs/canton/usermanual/kms/external_key_storage_aws.rst

..
   Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates.
..
   Proprietary code. All rights reserved.

.. _external_key_storage_aws:

> Configure External Key Storage and Usage with a AWS KMS

Reviewer: ```suggestion
Configure External Key Storage with AWS KMS
```

=======================================================

.. enterprise-only::

The following section describes the steps needed to enable :ref:`External Key Storage and Usage <kms_external_architecture>`
in Canton using AWS KMS. These steps include configuring AWS KMS, as well as, configuring this particular mode of operation.

.. note::
    Nodes configured with and without external KMS key storage can interact with each other,

>     including nodes using different KMS providers, as long as they follow the same crypto schemes.

Reviewer: This is true but `as long as they follow the same crypto schemes.` is a big enough caveat that we should emphasize it more. At least link it to the "Interoperability" section (see my other comment below)

In the example below, we configure **a new** Canton participant node (called ``participant1``) to generate and
store private keys in an external AWS KMS. The same configuration is applicable for all other node entities, e.g. domain-manager,
mediators, sequencers.

.. literalinclude:: /canton/includes/mirrored/enterprise/app/src/test/resources/aws-kms-provider-tagged.conf
   :language: none
   :start-after: user-manual-entry-begin: KmsProviderConfig
   :end-before: user-manual-entry-end: KmsProviderConfig

An example configuration that puts together both AWS KMS and external storage configuration is shown below:

.. literalinclude:: /canton/includes/mirrored/enterprise/app/src/test/resources/aws-kms-provider-tagged.conf
   :language: none
   :start-after: user-manual-entry-begin: AwsKmsProviderFullConfig
   :end-before: user-manual-entry-end: AwsKmsProviderFullConfig

By default nodes run a ``tink`` provider that is not compatible with KMS provider. If you want to continue to have other nodes
running a `non-KMS Canton` you are obliged to use a ``jce`` provider
and you must explicitly configure it to use the KMS supported algorithms as the required algorithms. Here is an

> example for a domain:

Reviewer: This paragraph assumes that the reader already has nodes running with KMS which is not necessarily the case. It's even at a bit at odds with the section which starts with
 
```
In the example below, we configure **a new** Canton participant node
```

This is of course important information but it's a bit confusing IMO that in the middle of the setup guide for AWS KMS on a participant I'm being told about tink being a default provider and that I need to configure jce on my domain. I'd put this in a separate sub-section, called something like "Interoperability with other nodes"

.. literalinclude:: /canton/includes/mirrored/enterprise/app/src/test/resources/aws-kms-provider-tagged.conf
   :language: none
   :start-after: user-manual-entry-begin: JceProviderDomainConfig
   :end-before: user-manual-entry-end: JceProviderDomainConfig

And here is an example for a participant:

.. literalinclude:: /canton/includes/mirrored/enterprise/app/src/test/resources/aws-kms-provider-tagged.conf
   :language: none
   :start-after: user-manual-entry-begin: JceProviderParticipantConfig
   :end-before: user-manual-entry-end: JceProviderParticipantConfig

In other words, a node running with a ``kms`` provider is only ever able to communicate with other nodes running
a ``kms`` or ``jce`` providers.

> A KMS only supports the :ref:`following cryptographic schemes <canton_supported_keys>`.

Reviewer: This links to a table detailing crypto schemes for all providers, not only KMS. So I'd rephrase to 

```suggestion
See :ref:`this table <canton_supported_keys>` for a description of cryptographic schemes supported by the KMS provider.
```

Setup with Pre-Generated Keys
-----------------------------

In the previous example, Canton creates its own keys on startup and initializes the identity of the nodes automatically.
If we want to use pre-generated keys, we need to manually initialize the identity of the nodes by

> adding the following flag in the config:

Reviewer: ```suggestion
adding the following to the node's configuration:
```

.. code-block:: none

    <node>.init.auto-init = false

This is applicable for ``participant`` and ``domain/domain-manager`` nodes.

Afterwards, we need register the keys in Canton by running the key registration command on each node.
For example for a ``participant`` we would run:

.. literalinclude:: /canton/includes/mirrored/enterprise/app/src/test/scala/com/digitalasset/canton/integration/tests/topology/TopologyManagementHelper.scala
   :language: scala
   :start-after: user-manual-entry-begin: ManualRegisterKmsKeys
   :end-before: user-manual-entry-end: ManualRegisterKmsKeys
   :dedent:

where `xyzKmsKeyId` is the AWS KMS identifier for a specific key (e.g. `KMS Key RN`). If we are using, for example,
`AWS cross account keys <https://docs.aws.amazon.com/kms/latest/developerguide/key-policy-modifying-external-accounts.html>`_
be aware that using the key id is not enough and we are required to register the key using its `ARN`.

> Here is a brief reminder of the initial keys that each node owns:

Reviewer: ```suggestion
Here is a reminder of the initial keys that each node owns:
```

you need to pre-generate the corresponding AWS KMS keys with the correct settings:

+-------------------+-----------------------------------------------------------------------+-----------------------------------------------------+
| **Provider**      | **SIGNING**                                                           | **ENCRYPTION**                                      |
+===================+=======================================================================+=====================================================+
| AWS               | - **Key Purpose:** `SIGN_VERIFY`                                      | - **Key Purpose:** `ENCRYPT_DECRYPT`                |
|                   | - **Key Algorithms:** `ECC_NIST_P256` or `ECC_NIST_P384`              | - **Key Algorithm:** `RSA_2048`                     |
+-------------------+-----------------------------------------------------------------------+-----------------------------------------------------+

Finally, using the previously registered keys, **we need to initialize our nodes**:

- :ref:`initialize domain <manually-init-domain>`
- :ref:`initialize participants <manually-init-participant>`

> Please be aware that for this specific use-case **we must use the keys that we registered and not generate new ones**.

Reviewer: I'd put this before the links to the manual init doc. Because people will click on the link and possibly miss this and copy paste the script that generate new keys. So something like

```
Below are links to guides on how to manually initialize participant and domain nodes. In those guides, keys are generated using console commands such as:

val identityKey =
  participant.keys.secret.generate_signing_key(name = participant.name + "-namespace")

Make sure to replace those commands with the ones shown above with which we registered our existing keys instead of generating new ones.
```

In other words, registering the keys replaces the `Keys Initialization` step in the guide to :ref:`manually initialize a node <manually_initializing_node>`.

.. _participant_aws_kms_migration:

> Participant Node Migration to KMS Crypto Provider

Reviewer: This section and the one after (manual key rotation) seem to be exactly the same for AWS and GCP. I think those we can abstract away to their own page to avoid repeating
